google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/JsonStreamWriter.java

```diff
@@ -108,6 +109,58 @@ public ApiFuture<AppendRowsResponse> append(JSONArray jsonArr)
     return append(jsonArr, -1);
   }
 
+  private void refreshWriter(TableSchema updatedSchema)
+      throws DescriptorValidationException, IOException {
+    Preconditions.checkNotNull(updatedSchema, "updatedSchema is null.");
+    LOG.info("Refresh internal writer due to schema update");
+    // Close the StreamWriter
+    this.streamWriter.close();
+    // Update JsonStreamWriter's TableSchema and Descriptor
+    this.tableSchema = updatedSchema;
+    this.descriptor =
+        BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(updatedSchema);
+    this.protoSchema = ProtoSchemaConverter.convert(this.descriptor);
+    this.totalMessageSize = protoSchema.getSerializedSize();
+    // Create a new underlying StreamWriter with the updated TableSchema and Descriptor
+    this.streamWriter = streamWriterBuilder.setWriterSchema(this.protoSchema).build();
+  }
+
+  private Message buildMessage(JSONObject json)
+      throws InterruptedException, DescriptorValidationException, IOException {
+    try {
```

Comment:
Can you change this function structure to

```suggestion
// Try four times, each time 10 seconds wait if not with the first retry
for (int i = 0; i < 5; i++) {
  try {
    JsonToProtoMessage.convertJsonToProtoMessage(
          this.descriptor, this.tableSchema, json, ignoreUnknownFields);
    break;
  } catch (Exceptions.JsonDataHasUnknownFieldException ex) {
    if (i == 4) {
      // No need to try last iteration
      break;
    }
    if (i > 0) {
      // Sleep 10 seconds if not with the first retry.
      sleep(10000);
    }
    GetWriteStreamRequest writeStreamRequest =
                  GetWriteStreamRequest.newBuilder()
              .setName(this.streamName)
              .setView(WriteStreamView.FULL)
              .build();
     WriteStream writeStream = client.getWriteStream(writeStreamRequest);
     refreshWriter(writeStream.getTableSchema());
  }
}
```

Reply:
Each GetWriteStream call will end up as a RPC to our backend and I don't want to add too much pressure to our servers in the case of a user side bogus field. Imagine, the user has 1K-100K streams with a bogus field constantly hitting us. Most of the resolvable case should be able to be addressed by the immediate GetWriteStream call, if that GetWriteStream call doesn't work out, I only want to retry once after a period that I know for sure the issue should be gone and error out. I want to retry as little as possible, in the error case, I don't mind the user get stuck.

Reply:
Can we change the iteration in my code from 5 to 3 in this case?

```suggestion
for (int i = 0; i < 3; i++) {
  try {
    JsonToProtoMessage.convertJsonToProtoMessage(
          this.descriptor, this.tableSchema, json, ignoreUnknownFields);
    break;
  } catch (Exceptions.JsonDataHasUnknownFieldException ex) {
    if (i == 3) {
      // No need to try last iteration
      break;
    }
    if (i == 1) {
      // Sleep 10 seconds if not with the first retry.
      sleep(30000);
    }
    GetWriteStreamRequest writeStreamRequest =
                  GetWriteStreamRequest.newBuilder()
              .setName(this.streamName)
              .setView(WriteStreamView.FULL)
              .build();
     WriteStream writeStream = client.getWriteStream(writeStreamRequest);
     refreshWriter(writeStream.getTableSchema());
  }
}

```

Reply:
No, I don't think it worth 3 times retry. I only want it to retry once.